File: promo-code-service/src/main/java/com/modsen/promocodeservice/exceptions/handler/GlobalExceptionHandler.java
package com.modsen.promocodeservice.exceptions.handler;

import com.modsen.promocodeservice.exceptions.PromoCodeAllReadyExistException;
import com.modsen.promocodeservice.exceptions.PromoCodeNotFoundException;
import com.modsen.promocodeservice.exceptions.error.AppError;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.util.HashMap;
import java.util.Map;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(PromoCodeNotFoundException.class)
    public ResponseEntity<AppError> handlerNotFoundException(RuntimeException ex) {
        String errorMessage = ex.getMessage();
        log.info("NotFoundException handler was called. Exception message : "+errorMessage);
        return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(new AppError(errorMessage));
    }

    @ExceptionHandler(PromoCodeAllReadyExistException.class)
    public ResponseEntity<AppError> handlerAlreadyExistException(RuntimeException ex) {
        String errorMessage = ex.getMessage();
        log.info("ConflictException handler was called. Exception message : "+errorMessage);
        return ResponseEntity.status(HttpStatus.CONFLICT)
                .body(new AppError(errorMessage));
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<Map<String, String>> handleValidationExceptions(
            MethodArgumentNotValidException ex) {

        Map<String, String> errors = new HashMap<>();

        ex.getBindingResult().getAllErrors().forEach(error -> {
            String fieldName = ((FieldError) error).getField();
            String errorMessage = error.getDefaultMessage();
            errors.put(fieldName, errorMessage);
        });
        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(errors);
    }
}

File: price-calculator-service/src/main/kotlin/org/example/pricecalculatorservice/exception/handler/GlobalExceptionHandler.kt
package org.example.pricecalculatorservice.exception.handler

import org.example.pricecalculatorservice.exception.NotFoundException
import org.example.pricecalculatorservice.exception.SomethingWrongTemperatureServiceException
import org.example.pricecalculatorservice.exception.appError.AppError
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class GlobalExceptionHandler {

    @ExceptionHandler(NotFoundException::class)
    fun handleNotFoundException(e: RuntimeException): ResponseEntity<AppError> {
        val message = e.message
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(message?.let { AppError(it) })
    }


    @ExceptionHandler(SomethingWrongTemperatureServiceException::class)
    fun handlerSomethingWrongTemperatureServiceException(e: RuntimeException): ResponseEntity<AppError> {
        val message = e.message
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(message?.let { AppError(it) })
    }
}